agent/csi/volumes.go

```diff
@@ -0,0 +1,263 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu     sync.RWMutex // To sync between Add(), Remove() and Reset() for "m" map
+	lock   sync.RWMutex // To sync between Get(), addVolumes() and removeVolumes()
+	m      map[string]*api.VolumeAssignment
+	plugin *NodePlugin
+}
+
+const maxRetry float64 = 25
+
+const initialBackoff = 100 * time.Millisecond
+const factor float64 = 2
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m:      make(map[string]*api.VolumeAssignment),
+		plugin: NewNodePlugin(),
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.plugin.GetPublisedPath(volumeID)
+}
+
+// Add adds one or more volumes to the volume map.
+func (r *volumes) Add(volumes ...api.VolumeAssignment) {
+	r.mu.Lock()
+	var volumeObjects []*api.VolumeAssignment
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(ctx).WithField("method", "(*volumes).Add").Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		volumeObjects = append(volumeObjects, v)
+	}
+	go r.addVolumes(volumeObjects)
+}
+
+func (r *volumes) addVolumes(volumeObjects []*api.VolumeAssignment) {
```

Comment:
The call to `Add` is no longer blocking, which is good. However, if one set of volumes is being processed by `addVolumes`, every subsequent set is going to be blocked waiting on that first set to finish.

You need to set this up so that a failing volume can't hold up other volumes. One solution may be to do something like this:

```go
// tryVolume is called in a goroutine
func (r *volumes) tryVolume(ctx context.Context, assignment *api.VolumeAssignment) {
	// nota bene: the Plugin object needs to contain a mutex. The CSI spec says
	// that the orchestrator is responsible for making sure there is no more
	// than one call in flight at any given time.
	if err := r.plugin.NodeStageVolume(ctx, assignment); err != nil {
		waitFor := initialBackoff
	retryStage:
		for i := 0; i < maxRetries; i++ {
			select {
			case <-ctx.Done():
				// selecting on ctx.Done() allows us to bail out of retrying early
				return
			case <-time.After(waitFor)
				// time.After is better than using time.Sleep, because it blocks
				// on a channel read, rather than suspending the whole
				// goroutine. That lets us do the above check on ctx.Done().
				//
				// time.After is convenient, but it has a key problem: the timer
				// is not garbage collected until the channel fires. this
				// shouldn't be a problem, unless the context is canceled, there
				// is a very long timer, and there are a lot of other goroutines
				// in the same situation.
				if err := r.plugin.NodeStageVolume(ctx, assignment); err == nil {
					break retryStage 
				}
			}
			// if the exponential factor is 2, you can avoid using floats by
			// doing bit shifts. each shift left increases the number by a power
			// of 2. we can do this because Duration is ultimately int64.
			waitFor := waitFor << 1
		}
	}
	
	// then, do something similar for publish
}
```

Each volume will be called in its own goroutine, and will not block any other volume from proceeding.